[Bernard] iP

bernarduskrishna opened the pull request:
# DukePro
>"Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

bernarduskrishna committed:
Finished Level-1: Greet, Echo, Exit

bernarduskrishna committed:
Finished Level-2: Add, List

bernarduskrishna committed:
Finished Level-3: Mark as Done

bernarduskrishna committed:
Finished Level-4: ToDo, Event, Deadline

bernarduskrishna committed:
Finished A-TextUiTesting

bernarduskrishna committed:
Kinda finished Level-5: Handle Errors. Not every errors are caught at the moment

bernarduskrishna committed:
Finished Level-6: Delete

bernarduskrishna committed:
Updated testcases

bernarduskrishna committed:
abstracted the outputs (a bit).

bernarduskrishna committed:
Finished A-Enums: Use Enums

bernarduskrishna committed:
Finished Level-7: Save

bernarduskrishna committed:
Finished Level-8: Dates and Times. Only did the minimal (input yyyy-mm-dd format, output MMM dd YYYY format) for both deadline and and event. Updated testcases.

bernarduskrishna committed:
Merge branch 'branch-Level-7'

bernarduskrishna committed:
Merge branch 'branch-Level-8'

bernarduskrishna committed:
deleted redundant code

bernarduskrishna committed:
Finished A-MoreOOP: Use More OOP

bernarduskrishna committed:
fixed bugs

bernarduskrishna committed:
Finished A-Packages: Organize into Packages

bernarduskrishna committed:
Finished A-JUnit: Add JUnit Tests. Added 3 tests for 3 different methods under 3 different classes.

bernarduskrishna committed:
Edited gitignore

bernarduskrishna committed:
Added Javadocs for all

bernarduskrishna committed:
Merge branch 'branch-A-JavaDoc'

bernarduskrishna committed:
Finished Level-9: Find

bernarduskrishna committed:
Edited testcases

bernarduskrishna committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/bern/functionalities/Ui.java

bernarduskrishna committed:
Change style for switch-case statement to comply with the standard

bernarduskrishna committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/bern/functionalities/Ui.java

s7u4rt99 commented on src/main/java/bern/functionalities/Parser.java:
```diff
@@ -0,0 +1,165 @@
+     * @param s The command from which a ToDo instance will be made.
+     * @return The ToDo instance.
+     */
+    public ToDo toDoFromString(String s) {
+        boolean isDone = s.substring(4, 5).equals("X");
+        String desc = s.substring(7);
+        ToDo ans = new ToDo(desc);
+        if (isDone) {
+            ans.markAsDone();
+        }
+        return ans;
+    }
+
+    /**
+     * A method to create a Deadline instance from a string command.
+     *
+     * @param s The command from which a Deadline instance will be made.
+     * @return The Deadline instance.
+     */
+    public Deadline deadlineFromString(String s) {
+        boolean isDone = s.substring(4, 5).equals("X");
+        String desc = s.substring(7, s.indexOf("(by: ") - 1);
+        String by = s.substring(s.indexOf("(by: ") + 5, s.length() - 1);
+        Deadline ans = new Deadline(desc, by);
+        if (isDone) {
+            ans.markAsDone();
+        }
+        return ans;
+    }
```
Really like the way you get a deadline and todo object from a string. Perhaps you can try to follow the naming conventions and call it getDeadlineFromString and getToDoFromString instead.

s7u4rt99 replied:
same for eventFromString and taskFromString

bryanwee023 replied:
It seems like`deadlineFromString()` is not instance specific. Perhaps tagging it as static makes more sense? Not really a coding style suggestion but thought it might be useful. The same could be applied for the other factory methods as well :)

s7u4rt99 commented on src/main/java/bern/functionalities/Ui.java:
```diff
@@ -0,0 +1,267 @@
+            result = "There is no matching task.";
+        }
+
+        return result;
```
Really like the way you broke the code up into few different parts to do different stuff, the structure is good and makes it easier to understand. Perhaps by adding javadocs and some comments in the body of the code would make it even easier to understand.

s7u4rt99 commented on src/main/java/bern/model/Task.java:
```diff
@@ -0,0 +1,55 @@
+        this.isDone = true;
+    }
+
+    public boolean findWord(String word) {
+        String[] words = description.split("\\s+");
+        for (String s : words) {
+            if (s.equals(word)) {
+                return true;
+            }
+        }
+        return false;
+    }
```
Perhaps you can add some javadocs here to follow the coding standard of this module

s7u4rt99 commented on src/main/java/bern/functionalities/Storage.java:
```diff
@@ -0,0 +1,72 @@
+        if (!file.exists()) {
+            return;
+        } else {
+            try {
+                Scanner s = new Scanner(file);
+                while (s.hasNext()) {
+                    String taskStr = s.nextLine();
+                    taskStr = taskStr.substring(taskStr.indexOf("["));
+                    Task temp = new Parser().taskFromString(taskStr);
+                    new TaskList().addTask(temp, arListTask);
+                }
+            } catch (FileNotFoundException e) {
+                System.out.println(e.getMessage());
+            }
```
Good and clean way to read the file to get the tasklist. Perhaps you can add some comments here to explain what each line does, so that it is more understandable.

s7u4rt99 commented on src/main/java/Bern.java:
```diff
@@ -0,0 +1,59 @@
+    public void run() {
+        Scanner myObj = new Scanner(System.in);
+
+        ArrayList<Task> arListTask = new ArrayList<>();
```
Like that you used an arraylist to store the tasks instead of an array due to arraylist being easily resizable. Perhaps you can call it arListTasks instead of arListTask to follow the naming convention.

s7u4rt99 reviewed (commented):
LGTM. Just some naming conventions, missing javadocs and comments nits to fix. Good job!

bernarduskrishna committed:
Fixed bugs

bernarduskrishna committed:
Add Bern class to the package as well

bernarduskrishna committed:
Merge branch 'add-gradle-support'

bernarduskrishna committed:
Finish A-Gradle: Use Gradle

bernarduskrishna committed:
Fix styling

bryanwee023 commented on src/main/java/bern/functionalities/Parser.java:
```diff
@@ -0,0 +1,165 @@
+     * @param input The command given.
+     * @return Boolean whether the command is of category "Done".
+     */
+    public boolean isDone(String input) {
```
I feel like creating a method for each keyword might gradually get messy over time. maybe consider a generalised alternative?

```Java
// Return whether first word of input matches the given keyword
public static boolean isKeyword(String input, String keyword) {
    String firstWord = input.split(" ", 2)[0];  //Get first word in input
    return firstWord.equals(keyword);
}
```

Again, not really a coding style violation, but I thought this could be useful.

bryanwee023 commented on src/main/java/bern/functionalities/Ui.java:
```diff
@@ -0,0 +1,276 @@
+     * @return The reply by the bot.
+     * @throws BernException If input is invalid.
+     */
+    public String ifToDo(String input, ArrayList<Task> arListTask) throws BernException {
```
The function name was a little misleading to me. Maybe consider a name that describes its purpose? (like `replyToDo()`)

I think I prefer this because its purpose of checking format is secondary as compared to its purpose of generating a reply.

bryanwee023 replied:
Can consider applying this to the other reply methods too!

bryanwee023 commented on src/main/java/bern/functionalities/Ui.java:
```diff
@@ -0,0 +1,276 @@
+     * @throws BernException If input is invalid.
+     */
+    public String getReply(Command c, String input, ArrayList<Task> arListTask) throws BernException {
+        switch (c) {
+        case DONE:
+            return new Ui().ifDone(input, arListTask);
+        case DEADLINE:
+            return ifDeadline(input, arListTask);
+        case EVENT:
+            return ifEvent(input, arListTask);
+        case TODO:
+            return ifToDo(input, arListTask);
+        case BYE:
+            return ifBye(input, arListTask);
+        case FIND:
+            return ifFind(input, arListTask);
+        case LIST:
+            return ifList(input, arListTask);
+        case DELETE:
+            return ifDelete(input, arListTask);
+        case INVALID:
+            throw new InvalidCommandException(input);
+        default:
+            return "";
```
This might just be my preference but I think long switch cases might make code less readable.
Maybe consider the usage of `HashTable<Command, Function>` ? Although this is probably down to individual preference.

bryanwee023 commented on src/main/java/bern/functionalities/Ui.java:
```diff
@@ -0,0 +1,276 @@
+     * @param arListTask The initial ArrayList of Tasks.
+     */
+    public void processInput(String input, ArrayList<Task> arListTask) {
+        try {
+            if (parser.isDone(input)) {
+                System.out.println(new Ui().getReply(Command.DONE, input, arListTask));
+                storage.writeIntoFile(arListTask);
+            } else if (parser.isDeadline(input)) {
+                System.out.println(new Ui().getReply(Command.DEADLINE, input, arListTask));
+                storage.writeIntoFile(arListTask);
+            } else if (parser.isEvent(input)) {
+                System.out.println(new Ui().getReply(Command.EVENT, input, arListTask));
+                storage.writeIntoFile(arListTask);
+            } else if (parser.isToDo(input)) {
+                System.out.println(new Ui().getReply(Command.TODO, input, arListTask));
+                storage.writeIntoFile(arListTask);
+            } else if (parser.isBye(input)) {
+                System.out.println(new Ui().getReply(Command.BYE, input, arListTask));
+            } else if (parser.isList(input)) {
+                System.out.println(new Ui().getReply(Command.LIST, input, arListTask));
+                storage.writeIntoFile(arListTask);
+            } else if (parser.isDelete(input)) {
+                System.out.println(new Ui().getReply(Command.DELETE, input, arListTask));
+                storage.writeIntoFile(arListTask);
+            } else if (parser.isFind(input)) {
+                System.out.println(new Ui().getReply(Command.FIND, input, arListTask));
+            } else {
+                new Ui().getReply(Command.INVALID, input, arListTask);
+            }
```
Constantly reinitialising Ui might seem redundant. Maybe make Ui methods static? I think that might make your code look cleaner :)

bryanwee023 commented on src/main/java/bern/functionalities/TaskList.java:
```diff
@@ -0,0 +1,30 @@
+     * @param task The additional task.
+     * @param arListTask The initial ArrayList of Tasks.
+     */
+    public void addTask(Task task, ArrayList<Task> arListTask) {
```
The name `arrListTask` suggests that the variable is a task. Consider `taskArrList` instead?

bryanwee023 reviewed (commented):
I like your coding style, looks really clean to me 👍 
Easy to read and nicely abstracted methods too :)

bernarduskrishna committed:
Do the bare minimum to get by for GUI.

bernarduskrishna committed:
Add fxml

bernarduskrishna committed:
Merge branch 'Level-10'

bernarduskrishna committed:
Add assertion

The code does not have any assertion.

This causes the code to be prone to errors and bad for long-term
maintenance.

Let's add some assert statements.

Most of the assertions are on string as this app is mainly a CLI
based app.

Unnecessarily long commit for such a simple task because this is
a requirement.

bernarduskrishna committed:
Decrease number of "magic numbers"

There are a lot of magic numbers.

This makes the logic of the code harder to understand.

Let's make the derivation of the "magic number" is made explicit.

Making it explicit makes the code easier to read by other people.

Unnecessarily long commit message because requirements.

bernarduskrishna committed:
Merge pull request #1 from bernarduskrishna/branch-A-Assertions

Add assertion

bernarduskrishna committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/bern/functionalities/Ui.java

bernarduskrishna committed:
There is no feature to detect duplicate

This might be a problem if someone forgot they have already added
a task.

Let's create a detect duplicate feature when someone is adding
a new task.

So that users won't have duplicate entries.

bernarduskrishna committed:
Merge branch 'C-DetectDuplicates'

bernarduskrishna committed:
Catch non-Bern Exceptions

bernarduskrishna committed:
Close platform upon "bye" and delete unnecessary files

bernarduskrishna committed:
Add Ui.png

bernarduskrishna committed:
change main to Launcher

bernarduskrishna committed:
Fix bug whereby "bye" when list is empty will cause subsequent
launch to throw an error

bernarduskrishna committed:
Fix bug whereby deadline/event can have tags other than /by or /at

bernarduskrishna committed:
Update README.md

bernarduskrishna committed:
Update README.md
